src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs

```diff
@@ -10,4 +10,513 @@
 --   they were constants or not and dispatches logic based on that
 module Juvix.Backends.Michelson.DSL.InstructionsEff where
 
+import qualified Data.Set as Set
+import qualified Juvix.Backends.Michelson.Compilation.Types as Types
+import qualified Juvix.Backends.Michelson.Compilation.VirtualStack as VStack
+import qualified Juvix.Backends.Michelson.DSL.Environment as Env
 import qualified Juvix.Backends.Michelson.DSL.Instructions as Instructions
+import qualified Juvix.Core.ErasedAnn.Types as Ann
+import qualified Juvix.Core.Usage as Usage
+import Juvix.Library hiding (abs, and, or, xor)
+import qualified Juvix.Library (abs)
+import qualified Michelson.Untyped.Instr as Instr
+import qualified Michelson.Untyped.Type as Untyped
+import qualified Michelson.Untyped.Value as V
+import Prelude (error)
+
+--------------------------------------------------------------------------------
+-- Main Functionality
+--------------------------------------------------------------------------------
+
+inst ∷ Env.Reduction m ⇒ Types.NewTerm → m Env.Expanded
+inst (t, _usage, ty) =
+  case t of
+    Ann.Var symbol → var symbol
+    Ann.LamM c a b → let v = (lambda c a b ty) in v <$ consVal v ty
+    Ann.AppM fun a → appM fun a
+    Ann.Prim prim' →
+      case prim' of
+        Types.Inst _ → constructPrim prim' ty
+        Types.Constant m → do
+          consVal (Env.Constant m) ty
+          pure (Env.Constant m)
+
+add,
+  mul,
+  sub,
+  ediv,
+  and,
+  xor,
+  neq,
+  eq,
+  lt,
+  gt,
+  le,
+  ge,
+  neg,
+  abs,
+  isNat,
+  or ∷
+    Env.Reduction m ⇒ Types.Type → [Types.NewTerm] → m Env.Expanded
+add = intGen Instructions.add (+)
+mul = intGen Instructions.mul (*)
+sub = intGen Instructions.sub (-)
+and = onBoolGen Instructions.and (&&)
+xor = onBoolGen Instructions.xor (/=)
+or = onBoolGen Instructions.or (||)
```

Comment:
For some of these, they are more polymorphic than what I've written here. They also operate on bytes and mutez...

Though mutez isn't in `V.Value'`?

Can these types not be computed at compile time?

Reply:
`V.Value'` has an `Int`, which can be `Int`, `Nat`, or `Mutez` in Michelson. Morley has all of those in `Type`.